### ISSUE TITLE
Using prediction heads

### ISSUE THREAD

**Issue by christian-storm** [explosion/spacy-transformers#61]
I'm probably being daft but is there an easy way to incorporate prediction heads?  For instance, the Bert Model with a language modeling head on top.  In other words, something akin to 
`model = BertForMaskedLM.from_pretrained('bert-base-uncased')`
in pytorch_transformers.

**Comment by honnibal** [comment]
It's a little more complicated because we're going for a design where you can have multiple "heads" attached, without recomputing the token vectors each time.

Currently the only head we have implemented is the text categorizer, but here's how it would work for something like the language model head:

1. Create a new class, `PyTT_LanguageModel`. It'll be similar to the `PyTT_TextCategorizer` in `spacy_pytorch_transformers/pipeline/textcat.py`
2. In the `PyTT_LanguageModel.Model` class method, we need to return a Thinc model. This part's arguably a bit obscure, and there are a couple of ways to do it -- see below.
3. Write the `predict()` and `set_annotations()` methods. You probably want to store the predictions in some extension attributes.

### Defining the model without Thinc

The approach that uses the least Thinc-specific stuff is to create your Torch model as normal, and then pass it into the `thinc.extra.wrappers.PyTorchWrapper` class. Here's a sketch of how the `Model()` and `predict()` methods might look (untested):

```python
import pytorch_transformers as pytt
import thinc.extra.wrappers

class PyTT_LanguageModel(spacy.pipeline.Pipe):
    @classmethod
    def Model(cls, **kwargs):
        config = pytt.BertConfig(kwargs["pytt_config"])
        pytorch_model = pytt.BertLMPredictionHead(pytt_config)
        return thinc.extra.wrappers.PyTorchWrapper(pytorch_model)

    def predict(self, docs):
        tensors = [doc._.pytt_last_hidden_state for doc in docs]
        # Here you could concatenate the tensors into one batch. For simplicity, I'll
        # just do one doc at a time
        tensors = [t.reshape((1,) + t.shape) for t in tensors]
        outputs = [self.model(t) for t in tensors]
        return outputs
```

I'm not immediately sure how to load pretrained weights for just the LM head from `pytorch_transformers`, without dragging in the BERT weights as well. Presumably there's a way? If you can load the weights once, then the `PyTorchWrapper` class's to/from disk/bytes methods should work. So, if that part turns out to be a bit hacky, at least you should only have to do it once --- if you're loading the model up during production, you won't need to be calling into the weight loading code from PyTorch Transformers.

### Chaining things together with Thinc

When we actually implement this, we'll probably want to define the model such that it takes a batch of `Doc` objects and returns the tensors directly. This is more like how the text classifier model works. We would have a function in the `model_registry.py` module that uses the `chain()` higher-order function to compose some functions together, so that we get the input and output how we want it. The `with_square_sequences` function in Thinc *should* do the batching correctly, although I haven't tested it thoroughly. We might need another function to pass forward masking or something, not sure.

```python

@register_model("pytt_lm")
def pytt_language_model(pytt_model):
    """Apply a PyTT language model head"""
    return chain(
        get_pytt_last_hidden,
        with_square_sequences(
            PyTorchWrapper(pytt_model))
    )
```

### Defining the actual model in Thinc

For the text classifier, I didn't bother using the PyTorch wrapper just for the output, as it's a simple feed-forward layer and we wouldn't need to load pretrained weights. For the LM, I think probably we'll want to stick with the wrapper. Defining the layer in Thinc would make it harder to get the initial weights loaded in correctly though, and getting the details of the objective might be too annoying.

**Comment by honnibal** [comment]
Just to clarify again: we definitely want to have the `PyTT_LanguageModel` pipeline component in the library. It's just not there yet, as our first priority was to get updated and address the serialization problems some users were reporting.

**Comment by christian-storm** [comment]
As always, thanks for the detailed and rapid response. 

I believe there is a way to load the LM head weights given BertForMaskedLM's __init__
    
    def __init__(self, config):
        super(BertForMaskedLM, self).__init__(config)
        self.bert = BertModel(config)
        self.cls = BertOnlyMLMHead(config)

        self.init_weights()
        self.tie_weights()

and apply them in the `forward` pass
   
    def forward(self, input_ids, attention_mask=None, token_type_ids=None, position_ids=None, head_mask=None, masked_lm_labels=None):

        outputs = self.bert(input_ids,
                            attention_mask=attention_mask,
                            token_type_ids=token_type_ids,
                            position_ids=position_ids, 
                            head_mask=head_mask)

        sequence_output = outputs[0]
        prediction_scores = self.cls(sequence_output)


I guess I wasn't too daft...not as straightforward as I hoped.  If I did implement it, I'm guessing it would be best to add it to spacy-pytorch-transformers?    There is no burning need on my end right now, I may hold off for now.

**Comment by enzoampil** [comment]
How can we have the above process automated into a CLI that takes in any pytorch model with a BERT + head architecture and turns it into a spacy model that can be integrated into a pipeline?

**Comment by svlandeg** [comment]
Since this post, the library has been significantly rewritten and published today as v. 1.0 and we'll publish the corresponding spaCy v. 3.0 very soon as well  🥳 

With [spaCy 3](https://nightly.spacy.io/usage/embeddings-transformers), it becomes much easier to use the transformer in your custom pipelines, share the weights across multiple classification tasks to do transfer / multi-task learning, and even combine all this with your [custom models and architectures](https://nightly.spacy.io/usage/layers-architectures).

We'll be moving the support to the spaCy repo, so I'll be closing this in the meantime. New bugs can be opened on the [spaCy GitHub tracker](https://github.com/explosion/spaCy/issues) and feature requests or ask for coding help can be posted on the [spaCy discussions forum](https://github.com/explosion/spaCy/discussions). 

Thanks all for testing the early prereleases of this package! All the feedback has been really helpful :-)
